crates/libs/implement/src/lib.rs

```diff
@@ -124,14 +124,14 @@ pub fn implement(attributes: proc_macro::TokenStream, original_type: proc_macro:
                     }
                 }
             }
-            fn AddRef(&mut self) -> u32 {
+            fn AddRef(&self) -> u32 {
                 self.count.add_ref()
             }
-            fn Release(&mut self) -> u32 {
+            unsafe fn Release(&self) -> u32 {
```

Comment:
Is it possible to make `Release` take `self`? If so, I think you could make this safe.

I'm guessing that doesn't fit with lots of COM patterns, since I remember making lots of manual calls to `Release` back when I've done COM programming before. Still, using this object after calling `Release` seems suspect, since `Release` might drop itself, but I guess that's why this is unsafe.

Reply:
`Release` gets called in the `Drop` impl of COM objects, and we only have a `&mut self` in `drop` so I don't think it's possible for `Release` to take `self`. This is unfortunately where COM's semantics and Rust semantics don't 100% align and so we're left with a dangling pointer for a short amount of time. I think the unfortunate answer is that we just have to be careful. Luckily, users will almost never interact with this method and instead `Release` will get called automatically on drop.